### PR TITLE
Allow using FPRoundingMode when see VK_KHR_16bit_storage

### PR DESCRIPTION
**Pull request by antiagainst**
According to the extension, FPRoundingMode should be allowed to
use without requiring Kernel capability when VK_KHR_16bit_storage
is enabled.

### CODE REVIEW

File: source/val/validation_state.cpp [hunk 1]
```diff
@@ -303,6 +304,12 @@ void ValidationState_t::RegisterCapability(SpvCapability cap) {
   }
 }
 
+void ValidationState_t::RegisterExtension(const char* extension) {
+  if (!strcmp(extension, "SPV_KHR_16bit_storage")) {
+    features_.free_fp_rounding_mode = true;
```
Comment (dneto0) [inline source/val/validation_state.cpp]: The flag should be flipped when we see the corresponding capabilities.
The reason is that a future version of an API can bring the extension into core, and then you'll have modules with the capabilities but not the OpExtension.

File: source/val/validation_state.h [hunk 2]
```diff
@@ -172,6 +175,12 @@ class ValidationState_t {
   /// Registers the capability and its dependent capabilities
   void RegisterCapability(SpvCapability cap);
 
+  // Registers the given extension.
```
Comment (dneto0) [inline source/val/validation_state.h]: I think you won't need this.

File: source/val/validation_state.cpp [hunk 3]
```diff
@@ -15,6 +15,7 @@
 #include "val/validation_state.h"
 
 #include <cassert>
+#include <cstring>
```
Comment (dneto0) [inline source/val/validation_state.cpp]: Won't need this.

File: source/validate_instruction.cpp [hunk 4]
```diff
@@ -79,11 +80,24 @@ CapabilitySet RequiredCapabilities(const AssemblyGrammar& grammar,
...
   spv_operand_desc operand_desc;
-  if (SPV_SUCCESS == grammar.lookupOperand(type, operand, &operand_desc)) {
-    return operand_desc->capabilities;
+  const auto ret = state.grammar().lookupOperand(type, operand, &operand_desc);
+  if (ret == SPV_SUCCESS) {
+    CapabilitySet result = operand_desc->capabilities;
+
+    // Allow FPRoundingMode decoration if requested
+    if (state.features().free_fp_rounding_mode &&
+        !strcmp(operand_desc->name, "FPRoundingMode")) {
```
Comment (dneto0) [inline source/validate_instruction.cpp]: Compare against the value not the string.

File: source/validate_instruction.cpp [hunk 5]
```diff
@@ -18,6 +18,7 @@
 
 #include <algorithm>
 #include <cassert>
+#include <cstring>
```
Comment (dneto0) [inline source/validate_instruction.cpp]: Won't need this.

File: source/validate_instruction.cpp [hunk 6]
```diff
@@ -331,6 +343,10 @@ spv_result_t RegisterDecorations(ValidationState_t& _,
 spv_result_t InstructionPass(ValidationState_t& _,
                              const spv_parsed_instruction_t* inst) {
   const SpvOp opcode = static_cast<SpvOp>(inst->opcode);
+  if (opcode == SpvOpExtension) {
```
Comment (dneto0) [inline source/validate_instruction.cpp]: Won't need this now.

File: test/val/val_data_test.cpp [hunk 7]
```diff
@@ -540,4 +541,46 @@ OpTypeForwardPointer %_ptr_Generic_struct_A Generic
   ASSERT_EQ(SPV_SUCCESS, ValidateInstructions());
 }
 
+TEST_F(ValidateData, ext_16bit_storage_allow_free_fp_rounding_mode) {
+  for (const char* ext : {"RTE", "RTZ", "RTP", "RTN"}) {
+    string str = string(R"(
+        OpCapability Shader
+        OpCapability Linkage
+        OpExtension "SPV_KHR_16bit_storage"
```
Comment (dneto0) [inline test/val/val_data_test.cpp]: Test against the capabilities, not the extension string.

File: source/val/validation_state.cpp (hunk 3 above)
Comment (antiagainst) [inline source/val/validation_state.cpp]: Done

File: source/val/validation_state.cpp (hunk 1 above)
Comment (antiagainst) [inline source/val/validation_state.cpp]: Yep. Changed.

File: source/val/validation_state.h (hunk 2 above)
Comment (antiagainst) [inline source/val/validation_state.h]: Removed.

File: source/validate_instruction.cpp (hunk 5 above)
Comment (antiagainst) [inline source/validate_instruction.cpp]: Removed

File: source/validate_instruction.cpp (hunk 4 above)
Comment (antiagainst) [inline source/validate_instruction.cpp]: That also requires to consider the `type`. Changed.

File: source/validate_instruction.cpp (hunk 6 above)
Comment (antiagainst) [inline source/validate_instruction.cpp]: Done

File: test/val/val_data_test.cpp (hunk 7 above)
Comment (antiagainst) [inline test/val/val_data_test.cpp]: Done

**Comment by dneto0** [comment]
Rebased and pushed into master as 40be4bdd34a0fce1dfcc97f571704e7980d09a36

**Pull request closed without merging**